src/server.zig

```diff
@@ -53,79 +53,108 @@ pub fn listenAndServe(
-        comptime handler: RequestHandler(@TypeOf(context)),
-    ) !void {
-        var stream = net.StreamServer.init(.{ .reuse_address = true });
-        defer stream.deinit();
+        context: Context,
+
+        stream: net.StreamServer,
 
         // client queue to clean up clients after connection is broken/finished
-        const Client = ClientFn(@TypeOf(context), handler);
-        var clients = Queue(*Client).init();
-
-        // Force clean up any remaining clients that are still connected
-        // if an error occured
-        defer while (clients.get()) |node| {
-            const data = node.data;
-            data.stream.close();
-            gpa.destroy(data);
-        };
+        clients: Queue(*Client),
+
+        /// Initializes a new `Server` instance
+        pub fn init(
+            /// Memory allocator, for general usage.
+            /// Will be used to setup an arena to free any request/response data.
+            gpa: *Allocator,
+            /// Address the server is listening at
+            address: net.Address,
+            /// Runtime context allowing to pass data between handlers.
+            /// Thread-safety must be guaranteed by the caller.
+            context: Context,
+        ) Self {
+            return Self{
+                .should_quit = atomic.Atomic(bool).init(false),
+                .gpa = gpa,
+                .address = address,
+                .context = context,
+                .stream = net.StreamServer.init(.{ .reuse_address = true }),
+                .clients = Queue(*Client).init(),
+            };
+        }
 
-        try stream.listen(address);
+        pub fn deinit(self: *Self) void {
+            self.shutdown();
+            // Force clean up any remaining clients that are still connected
+            // if an error occured
+            while (self.clients.get()) |node| {
+                const data = node.data;
+                data.stream.close();
+                self.gpa.destroy(data);
+            }
+            self.stream.deinit();
+            self.* = undefined;
+        }
 
-        while (!self.should_quit.load(.SeqCst)) {
-            var connection = stream.accept() catch |err| switch (err) {
-                error.ConnectionResetByPeer, error.ConnectionAborted => {
-                    log.err("Could not accept connection: '{s}'", .{@errorName(err)});
-                    continue;
-                },
-                else => return err,
-            };
+        pub fn start(self: *Self) !void {
+            try self.stream.listen(self.address);
+        }
 
-            // setup client connection and handle it
-            const client = try gpa.create(Client);
-            client.* = Client{
-                .stream = connection.stream,
-                .node = .{ .data = client },
-                .frame = async client.run(gpa, &clients, context),
-            };
+        pub fn getContext(self: *Self) !*Client {
```

Comment:
I find the name to be confusing. The user is asking for a context, but instead receives a client, where only the `stream` is useful for them. It at the very least requires doc comments.